src/type/sum.rs

```diff
@@ -0,0 +1,367 @@
+// Copyright 2019
+//     by  Centrality Investments Ltd.
+//     and Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::tm_std::*;
+
+use crate::{
+	form::{CompactForm, Form, MetaForm},
+	IntoCompact, NamedField, Registry, UnnamedField,
+};
+use derive_more::From;
+use serde::Serialize;
+
+/// A Sum type (consisting of variants) e.g an enum
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, Serialize, From)]
+#[serde(bound = "F::Type: Serialize")]
+#[serde(rename_all = "lowercase")]
+pub enum TypeSum<F: Form = MetaForm> {
+	// todo: [AJ] potentially extract struct TypeSumVariants<Variant>
+	/// A C-like enum type.
+	ClikeEnum(TypeSumClikeEnum<F>),
+	/// A Rust enum, aka tagged union.
+	Enum(TypeSumEnum<F>),
```

Comment:
Should be unified same as in product types. A builder checks that properties are hold. Results in better serialization.